## Задание 2

In [11]:
import numpy as np
from sklearn import datasets
from sklearn.model_selection import cross_val_score
from sklearn.naive_bayes import BernoulliNB, MultinomialNB, GaussianNB
import matplotlib.pyplot as plt

%matplotlib inline

Загрузим датасеты, выведем несколько строчек из обучающих выборок и названия признаков (где это возможно).

In [2]:
digits = datasets.load_digits()
print(digits.data[:5])

[[  0.   0.   5.  13.   9.   1.   0.   0.   0.   0.  13.  15.  10.  15.
    5.   0.   0.   3.  15.   2.   0.  11.   8.   0.   0.   4.  12.   0.
    0.   8.   8.   0.   0.   5.   8.   0.   0.   9.   8.   0.   0.   4.
   11.   0.   1.  12.   7.   0.   0.   2.  14.   5.  10.  12.   0.   0.
    0.   0.   6.  13.  10.   0.   0.   0.]
 [  0.   0.   0.  12.  13.   5.   0.   0.   0.   0.   0.  11.  16.   9.
    0.   0.   0.   0.   3.  15.  16.   6.   0.   0.   0.   7.  15.  16.
   16.   2.   0.   0.   0.   0.   1.  16.  16.   3.   0.   0.   0.   0.
    1.  16.  16.   6.   0.   0.   0.   0.   1.  16.  16.   6.   0.   0.
    0.   0.   0.  11.  16.  10.   0.   0.]
 [  0.   0.   0.   4.  15.  12.   0.   0.   0.   0.   3.  16.  15.  14.
    0.   0.   0.   0.   8.  13.   8.  16.   0.   0.   0.   0.   1.   6.
   15.  11.   0.   0.   0.   1.   8.  13.  15.   1.   0.   0.   0.   9.
   16.  16.   5.   0.   0.   0.   0.   3.  13.  16.  16.  11.   5.   0.
    0.   0.   0.   3.  11.  16.   9.   0.]
 [  0. 

In [3]:
breast_cancer = datasets.load_breast_cancer()
print(breast_cancer.feature_names)
print(breast_cancer.data[:5])

['mean radius' 'mean texture' 'mean perimeter' 'mean area'
 'mean smoothness' 'mean compactness' 'mean concavity'
 'mean concave points' 'mean symmetry' 'mean fractal dimension'
 'radius error' 'texture error' 'perimeter error' 'area error'
 'smoothness error' 'compactness error' 'concavity error'
 'concave points error' 'symmetry error' 'fractal dimension error'
 'worst radius' 'worst texture' 'worst perimeter' 'worst area'
 'worst smoothness' 'worst compactness' 'worst concavity'
 'worst concave points' 'worst symmetry' 'worst fractal dimension']
[[  1.79900000e+01   1.03800000e+01   1.22800000e+02   1.00100000e+03
    1.18400000e-01   2.77600000e-01   3.00100000e-01   1.47100000e-01
    2.41900000e-01   7.87100000e-02   1.09500000e+00   9.05300000e-01
    8.58900000e+00   1.53400000e+02   6.39900000e-03   4.90400000e-02
    5.37300000e-02   1.58700000e-02   3.00300000e-02   6.19300000e-03
    2.53800000e+01   1.73300000e+01   1.84600000e+02   2.01900000e+03
    1.62200000e-01   6.65

Видим, что в случае digits --- **целочисленные неотрицательные признаки**, а в случае breast_cancer --- **вещественные**.

Сравним качество работы наивных байесовских оценок на датасете **digits**.

In [21]:
def score(X, y):
    bnb = BernoulliNB()
    mnb = MultinomialNB()
    gnb = GaussianNB()
    print('BernoulliNB: {:.2f}'.format(np.mean(cross_val_score(bnb, X, y))))
    print('MultinomialNB: {:.2f}'.format(np.mean(cross_val_score(mnb, X, y))))
    print('GaussianNB: {:.2f}'.format(np.mean(cross_val_score(gnb, X, y))))

In [26]:
X = digits.data
y = digits.target
score(X, y)

BernoulliNB: 0.83
MultinomialNB: 0.87
GaussianNB: 0.82


Теперь сравним качество работы наивных байесовских оценок на датасете **breast_cancer**.

In [25]:
X = breast_cancer.data
y = breast_cancer.target
score(X, y)

BernoulliNB: 0.63
MultinomialNB: 0.89
GaussianNB: 0.94


1. На датасете digits максимальное качество классификации у **MultinomialNB: 0.87**.
2. На датасете breast_cancer максимальное качество классификации у **GaussianNB: 0.94**.
3. Верны следующие утверждения:
c) мультиномиальное распределение лучше показало себя на выборке с целыми неотрицательными значениями признаков,
d) на вещественных признаках лучше всего сработало нормальное распределение